src/win/fs.c

```diff
@@ -144,26 +144,96 @@ void uv__fs_init(void) {
 }
 
 
+static int32_t fs__decode_wtf8_char(const char** input) {
```

Comment:
This function is very similar to `uv__utf8_decode1()` in src/idna.c. It'd be nicer to try to share code with it.

<details>

https://github.com/libuv/libuv/blob/e7ecd116e0e013bf8c01d66505ced4ce1b58c005/src/idna.c#L26-L103

</details>

One difference is that src/idna.c rejects surrogate pairs. This function seems to allow them (on line 165-166)?

Reply:
Hm, interesting. Yeah, it does seem like we could use this for decoding either case, then for `uv__utf8_decode1` add a check for that extra surrogate pair case. I think that may be the crux of the difference between UTF-8 (which requires them to be a decoding error) and WTF-8 (which requires them to round-trip correctly from UCS-2 / potentially ill-formed UTF-16): https://simonsapin.github.io/wtf-8/#surrogates-byte-sequences

I was considering a followup PR that exported these encoding/decoding utilities, as potentially useful helper functions for anyone building with libuv. I could merge them then, if that sounds good to you?

Reply:
Yes, this is the only difference between WTF-8 and UTF-8: WTF-8 simply doesn't reject code points in the range U+D800 through U+DFFF, whereas UTF-8 explicitly checks for and rejects those since they are not valid standalone code points and are supposed to only be used in pairs by UTF-16 for encoding large code points. Thus WTF-8 is actually simpler to validate since it skips that check. Both encodings still check for and reject other ways of being invalid, including simply being malformed, encoding a code point higher than U+10FFFF, and overlong encodings.

Reply:
In general, I think that whether to error on surrogates can be an option to both the encoding and decoding functions. Or alternatively, just use WTF-8 everywhere, which honestly is a very reasonable choice—there's not a lot of cases where it seems like libuv's job to make sure that stray surrogates don't happen, especially since WTF-8 is *exactly* the encoding that allows arbitrary data interpreted as UTF-16 to be round-tripped, but I don't really care to fight that fight beyond the file system API where we're forced to deal with it.